sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -13,93 +13,86 @@ namespace Azure.Core
 
         /// <summary>
         /// Gets the <see cref="Response"/> associated with this message. Throws an exception if it wasn't set yet.
         /// To avoid the exception use <see cref="HasResponse"/> property to check.
         /// </summary>
-        public Response Response
+        public new Response Response
         {
             get
             {
-                if (_response == null)
+                if (base.Response is null)
                 {
 #pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
-                    throw new InvalidOperationException("Response was not set, make sure SendAsync was called");
+                    throw new InvalidOperationException($"{nameof(Response)} is not set on this message. "
+                        + "This is may be because the message was not sent via pipeline.Send, "
+                        + "the pipeline transport did not populate the response, or because "
+                        + $"{nameof(ExtractResponse)} was called. You can check the {nameof(HasResponse)} "
+                        + "property to test whether the message has a response value before "
+                        + $"accessing the {nameof(Response)} property.");
 #pragma warning restore CA1065 // Do not raise exceptions in unexpected locations
                 }
-                return _response;
+                return (Response)base.Response;
             }
-            set => _response = value;
+
+            set => base.Response = value;
         }
 
         /// <summary>
         /// Gets the value indicating if the response is set on this message.
         /// </summary>
-        public bool HasResponse => _response != null;
+        public bool HasResponse => base.Response is not null;
 
-        internal void ClearResponse() => _response = null;
-
-        /// <summary>
-        /// The <see cref="System.Threading.CancellationToken"/> to be used during the <see cref="HttpMessage"/> processing.
-        /// </summary>
-        public CancellationToken CancellationToken { get; internal set; }
+        internal void ClearResponse() => Response = null!;
 
         /// <summary>
         /// The <see cref="ResponseClassifier"/> instance to use for response classification during pipeline invocation.
         /// </summary>
-        public ResponseClassifier ResponseClassifier { get; set; }
-
-        /// <summary>
-        /// Gets or sets the value indicating if response would be buffered as part of the pipeline. Defaults to true.
-        /// </summary>
-        public bool BufferResponse { get; set; }
+        public new ResponseClassifier ResponseClassifier
+        {
+            get => base.ResponseClassifier switch
+            {
+                ResponseClassifier responseClassifier => responseClassifier,
+                PipelineMessageClassifier messageClassifier => new ResponseClassifier.PipelineMessageClassifierAdapter(messageClassifier)
```

Comment:
Do we want to allocate a new wrapper on each call, or should we consider caching the wrapper and just clearning if `set` is called?

Reply:
I think it's cacheable -- the trade-off is adding a field to the message type which is allocated on every client service method call for what is a very edge scenario.  I didn't think the perf impact of the allocation for the edge case was that bad, since in typical operation, this is only called once by the transport.  I'd lean toward keeping it this way given that, but I'll give it another think before completing the PR.

Reply:
I don't think we need to cache it. This adapter is just to preserve compat. It should almost never happen in real life that an Azure pipeline end up having non-azure classifier.